akka-http2-support/src/test/scala/akka/http/impl/engine/http2/RequestParsingSpec.scala

```diff
@@ -0,0 +1,540 @@
+        }
+        val headerParser = HttpHeaderParser(parserSettings, log)
+        RequestParsing.parseRequest(headerParser, serverSettings, attributes)
+      }
+      parseRequest(subStream)
+    }
+
+    def shouldThrowMalformedRequest[T](block: ⇒ T): Exception = {
+      val thrown = the[RuntimeException] thrownBy block
+      thrown.getMessage should startWith("Malformed request: ")
+      thrown
+    }
+
+    "follow RFC7540" should {
+
+      // 8.1.2.1.  Pseudo-Header Fields
+
+      // All pseudo-header fields MUST appear in the header block before
+      // regular header fields.  Any request or response that contains a
+      // pseudo-header field that appears in a header block after a regular
+      // header field MUST be treated as malformed...
+
+      "not accept pseudo-header fields after regular headers" in pendingUntilFixed {
+        val pseudoHeaders = Vector(
+          ":method" → "GET",
+          ":scheme" → "https",
+          ":path" → "/"
+        )
+        forAll(0 until pseudoHeaders.length) { insertPoint: Int ⇒
+          // Insert the Foo header so it occurs before at least one pseudo-header
+          val (before, after) = pseudoHeaders.splitAt(insertPoint)
+          val modified = before ++ Vector("Foo" → "bar") ++ after
+          shouldThrowMalformedRequest(parse(modified))
+        }
+      }
+
+      // 8.1.2.2.  Connection-Specific Header Fields
+
+      // ...any message containing connection-specific header fields MUST
+      // be treated as malformed...
+
+      "not accept connection-specific headers" in pendingUntilFixed {
+        shouldThrowMalformedRequest {
+          // Add Connection header to indicate that Foo is a connection-specific header
+          parse(Vector(
+            ":method" → "GET",
+            ":scheme" → "https",
+            ":path" → "/",
+            "Connection" → "foo",
+            "Foo" → "bar"
+          ))
+        }
+      }
+
+      // 8.1.2.3.  Request Pseudo-Header Fields
+
+      // The ":method" pseudo-header field includes the HTTP method
+      // ...
+
+      "parse the ':method' pseudo-header correctly" in {
+        val methods = Seq("GET", "POST", "DELETE", "OPTIONS")
+        forAll(methods) { method: String ⇒
+          val request: HttpRequest = parse(
+            keyValuePairs = Vector(
+              ":method" → method,
+              ":scheme" → "https",
+              ":path" → "/"
+            ))
+          request.method.value should ===(method)
+        }
+      }
+
+      // The ":scheme" pseudo-header field includes the scheme portion of
+      // the target URI ([RFC3986], Section 3.1).
+      //
+      // ":scheme" is not restricted to "http" and "https" schemed URIs.  A
+      // proxy or gateway can translate requests for non-HTTP schemes,
+      // enabling the use of HTTP to interact with non-HTTP services.
+
+      "parse the ':scheme' pseudo-header correctly" in {
+        val schemes = Seq("http", "https")
```

Comment:
include a "banana" scheme to test that part of the spec that it allows other schemes too?

Reply:
`HttpRequest` checks its scheme when it's constructed. It can only have a scheme of `http`, `https`, `ws` or `wss`—no `banana` unfortunately. I was previously testing with `ws`/`wss` too, but I removed the tests for some reason. I'll try adding tests for `ws` and `wss` back in…

Reply:
Oh I see. That's fine then, I doubt anyone will come and request this specific thing any time soon anyway, adding ws / wss sounds good :+